internal/app.go

```diff
@@ -0,0 +1,187 @@
+package internal
+
+import (
+	"embed"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	ics "github.com/arran4/golang-ical"
+	"github.com/gin-gonic/gin"
+	"io"
+	"log"
+	"net/http"
+	"regexp"
+	"strings"
+)
+
+const src = "https://campus.tum.de/tumonlinej/ws/termin/ical?pStud=%s&pToken=%s"
+
+//go:embed courses.json
+var coursesJson string
+
+//go:embed buildings.json
+var buildingsJson string
+
+//go:embed static
+var static embed.FS
+
+type App struct {
+	engine *gin.Engine
+
+	courseReplacements   map[string]string
+	buildingReplacements map[string]string
+}
+
+func (a *App) Run() error {
+	err := json.Unmarshal([]byte(coursesJson), &a.courseReplacements)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal([]byte(buildingsJson), &a.buildingReplacements)
+	if err != nil {
+		return err
+	}
+	gin.SetMode("release")
+	a.engine = gin.New()
+	a.engine.Use(gin.Logger(), gin.Recovery())
+	a.configRoutes()
+	return a.engine.Run()
+}
+
+func (a *App) configRoutes() {
+	a.engine.Any("/", a.handleIcal)
+	f := http.FS(static)
+	a.engine.StaticFS("/files/", f)
+	a.engine.NoMethod(func(c *gin.Context) {
+		c.AbortWithStatus(http.StatusNotImplemented)
+	})
+}
+
+func (a *App) handleIcal(c *gin.Context) {
+	stud := c.Query("pStud")
+	token := c.Query("pToken")
+	if stud == "" || token == "" {
+		f, err := static.Open("static/index.html")
+		if err != nil {
+			c.AbortWithStatusJSON(http.StatusInternalServerError, err)
+			return
+		}
+		io.Copy(c.Writer, f)
+		return
+	}
+	resp, err := http.Get(fmt.Sprintf(src, stud, token))
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, err)
+		return
+	}
+	all, err := io.ReadAll(resp.Body)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, err)
+		return
+	}
+	cal, err := ics.ParseCalendar(strings.NewReader(string(all)))
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, err)
+		return
+	}
+	hasLecture := make(map[string]bool)
+	var newComponents []ics.Component
+	for _, component := range cal.Components {
```

Comment:
Would be nice to have a comment here what this block of code does, so folks don't have to parse the entire for/switch case story in order to understand whats going on.